docs/rfcs/0041-backend-native-api.md

```diff
@@ -0,0 +1,158 @@
+
+First, we will do at least three syscalls in every' read' operation: `open`, `seek`, `read`.
+
+```rust
+let mut f = fs::OpenOptions::new()
+    .read(true)
+    .open(&path)
+    .await
+    .map_err(|e| parse_io_error(&e, &path))?;
+
+if let Some(offset) = args.offset {
+    f.seek(SeekFrom::Start(offset))
+        .await
+        .map_err(|e| parse_io_error(&e, &path))?;
+};
+```
+
+To make everything worse, our `SeelableReader` is designed for object storage systems and works much slower on local fs.
+
+`SeelableReader` will try to prefetch data in memory and maintain an internal pointer to implement `AsyncSeek`. However, due to our poor (nearly no) optimization, `SeekableReader` is much slower than using a system call `seek`.
+
+## Ergonomic unfriendly
+
+`opendal` is not easy to use either.
+
+In our early adoption project `databend`, we can see a lot of code looks like:
+
+```rust
+let data_accessor = self.data_accessor.clone();
+let path = self.path.clone();
+let reader = SeekableReader::new(data_accessor, path.as_str(), stream_len);
+let reader = BufReader::with_capacity(read_buffer_size as usize, reader);
+Self::read_column(reader, &col_meta, data_type.clone(), arrow_type.clone()).await
+```
+
+And
+
+```rust
+op.stat(&path).run().await
+```
+
+## Conclusion
+
+So in this proposal, I expect to address those problems. After implementing this proposal, we have a faster and easier-to-use `opendal`.
+
+# Guide-level explanation
+
+To operate on an object, we will use `Operator::object()` to create a new handler:
+
+```rust
+let o = op.object("path/to/file");
+```
+
+All operations that are available for `Object` for now includes:
+
+- `stat`: get object metadata (return an error if not exist).
+- `delete`: delete an object.
+- `new_reader`: create a new reader to read data from this object.
+- `new_writer`: create a new writer to write data into this object.
+
+Here is an example:
+
+```rust
+use anyhow::Result;
+use futures::AsyncReadExt;
+
+use opendal::services::fs;
+use opendal::Operator;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let op = Operator::new(fs::Backend::build().root("/tmp").finish().await?);
+
+    let o = op.object("test_file");
+
+    // Write data info file;
+    let w = o.new_writer();
+    let n = w
+        .write_bytes("Hello, World!".to_string().into_bytes())
+        .await?;
+    assert_eq!(n, 13);
+
+    // Read data from file;
+    let mut r = o.new_reader();
+    let mut buf = vec![];
+    let n = r.read_to_end(&mut buf).await?;
+    assert_eq!(n, 13);
+    assert_eq!(String::from_utf8_lossy(&buf), "Hello, World!");
+
+    // Get file's Metadata
+    let meta = o.stat().await?;
+    assert_eq!(meta.content_length(), 13);
+
+    // Delete file.
+    o.delete().await?;
+
+    Ok(())
+}
+```
+
+# Reference-level explanation
+
+To support native `read` and `seek` operation on `fs`, we will split the `Read` operation into `sequential_read` and `random_read`:
```

Comment:
are these concepts identical to `AysncSeek` and `AsyncRead`?

Reply:
No, they are different.

- `sequential_read` intends to return a `Box<dyn AsyncRead>`, which will optimize for sequential read, like `read_all`.
- `random_read` intends to return a `Box<dyn AsyncRead + AsyncSeek>` which will optimize for random read, like required by parquet, seek to pos X, read Y, and seek to pos Z again and again.

Reply:
> they are only cared about by the backend implementor.

For opendal, the implementors should be taken into account as well, imo.  

Wrapping a bare `AsyncReader` with a `BufReader` seems to be the more common way. i.e. the implementors only care about the implementations of `AsyncRead + AsyncSeek`, let the `BufReader` do the optimization of `sequential_read`.

Reply:
> Wrapping a bare `AsyncReader` with a `BufReader` seems to be the more common way. i.e., the implementors only care about the implementations of `AsyncRead + AsyncSeek`, let the `BufReader` do the optimization of `sequential_read`.

Yep, we did this separation to make implementors easier.

- For backend supports `AsyncRead + AsyncSeek` natively like `fs`, we will return the underlying `RawFd` for `random_read`.
- For the backend doesn't support `AsyncSeek`, implementors need to implement `sequential_read` and return a `SeekableReader` for `random_read`. We have `sequential_read` here because we don't want every backend to implement its own `SeekableReader`.

Reply:
Does it a good idea to wrap with `BufReader` in `Object::reader()` so that both implementors and caller don't need to care about it? I can expose a `buf_limit` to the caller so that they can tune the buffer size.

Reply:
> we don't want every backend to implement its own SeekableReader.

That's pretty. 

Put it another way: do we need the "backend"s to implement method "sequential_read" since given a "SeekableReader", we(opendal as the "framework") can always wrap it with `BufReader`.

Reply:
> Does it a good idea to wrap with `BufReader` in `Object::reader()` so that both implementors and caller don't need to care about it? I can expose a `buf_limit` to the caller so that they can tune the buffer size.

> both implementors and caller don't need to care about it?

for the caller, not sure ... 

Since `BufReader` is not always better than the un-buffered read, at least for one-shot read, with known size, an ordinary read might be good enough (or even better than a buffered read, which read in extra data unnecessarily).

Reply:
Let's keep moving forward,  and refine the design step by step.